From: Aron Xu <aron@debian.org>
Date: Mon, 18 Jun 2012 17:11:40 +0800
Subject: no display desktop

---
 data/fcitx.desktop.in |    1 +
 1 file changed, 1 insertion(+)

diff --git a/data/fcitx.desktop.in b/data/fcitx.desktop.in
index 0429b5c..474ad47 100644
--- a/data/fcitx.desktop.in
+++ b/data/fcitx.desktop.in
@@ -4,6 +4,7 @@ _Comment=Input Method
 Exec=fcitx
 Icon=fcitx
 Terminal=false
+NoDisplay=true
 Type=Application
 Categories=System;Utility;
 StartupNotify=false
-- 
